src/mame/igs/igs_m027.cpp

```diff
@@ -114,9 +146,27 @@ void igs_m027_state::machine_start()
 {
 	m_dsw_io_select = 7;
 	m_unk2_write_count = 0;
+	m_runtime_decrypt_done = 0;
 
 	save_item(NAME(m_dsw_io_select));
 	save_item(NAME(m_unk2_write_count));
+	save_item(NAME(m_xortable));
+	save_item(NAME(m_runtime_decrypt_done));
+}
+
+u8 igs_m027_state::xor_read(offs_t offset, u8 data)
+{
+	return m_xortable[offset];
+}
+
+void igs_m027_state::xor_write(offs_t offset, u8 data)
+{
+	m_xortable[offset] = data;
+
+	if (offset == 0xff) // might not be the trigger
+	{
+		do_decrypt();
+	}
 }
```

Comment:
The current implementation of this will produce unexpected behaviour with saved states:

If you start the system, let it upload an XOR table, then load a saved state where the XOR table wasn’t decrypted, it won’t restore the initial empty state of the decrypted `user1` region.

You only re-run the decryption on the last byte of the XOR table being written, but on loading a saved state, you’ll run the decryption with whatever the immediate contents of the table is.  If the game writes to some earlier position in the table after writing the final byte, you’ll end up with different decrypted content to what you had when the saved state was created.

To avoid the issues, it might be easier to just use `.readonly().share(...)` for the `user1` ROM in the address map, use an object finder to get a pointer to it, and pass that to the decryption function.  That way the memory system will just save and restore the decrypted content for you and you won’t get these kinds of issues.